### PR TITLE
🎯 [SFT] add token accuracy metric

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

Adds a `token_accuracy` metric for training and evaluation in the SFTTrainer

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -546,3 +547,59 @@ def create_model_card(
...
+        loss = super().training_step(model, inputs)
+
+        # Compute token accuracy for training
+        if "labels" in inputs:
+            with torch.no_grad():
+                outputs = model(**inputs)
+                shift_logits = outputs.logits[..., :-1, :].contiguous()
+                shift_labels = inputs["labels"][..., 1:].contiguous()
+                train_accuracy = compute_token_accuracy(shift_logits, shift_labels)
+                self.log({"train_mean_token_accuracy": train_accuracy})
```
Comment (qgallouedec) [inline trl/trainer/sft_trainer.py:565]: Dors it work with multi gpu?

Comment (kashif) [reply trl/trainer/sft_trainer.py]: i moved it to `gather_for_metrics`

**Comment by kashif** [comment]
@lewtun it is now logging to the terminal nicely too:
```
{'loss': 14.3862, 'grad_norm': 40.221492767333984, 'learning_rate': 1.995675675675676e-05, 'mean_token_accuracy': tensor(0.5978, device='mps:0'), 'epoch': 0.0}
  0%|▏                                                           | 2/925 [00:39<4:49:34, 18.82s/itTrainer is attempting to log a value of "0.6009876132011414" of type <class 'torch.Tensor'> for key "eval/mean_token_accuracy" as a scalar. This invocation of Tensorboard's writer.add_scalar() is incorrect so we dropped this attribute.
{'eval_loss': 1.7303236722946167, 'eval_runtime': 74.6098, 'eval_samples_per_second': 2.56, 'eval_steps_per_second': 0.322, 'eval_mean_token_accuracy': tensor(0.6010, device='mps:0'), 'epoch': 0.0} 
```

**Review by qgallouedec — APPROVED** [review]
it shines ✨

File: trl/trainer/utils.py [hunk 2]
```diff
@@ -1647,3 +1647,24 @@ def flush_left(mask: torch.Tensor, *tensors: torch.Tensor) -> tuple[torch.Tensor
...
+    Compute the mean token accuracy.
+    """
+    # Get predictions
+    predictions = logits.argmax(dim=-1)
+
+    # Create mask for non-padding tokens (assuming pad_token_id is ignore_index)
+    mask = labels != ignore_index
+
+    # Calculate accuracy only on non-padding tokens
+    correct_predictions = (predictions == labels) & mask
```
[the remark this answers is not in the document]
Comment (kashif) [reply trl/trainer/utils.py:1663]: checking
